src/middlewared/middlewared/plugins/acm.py

```diff
@@ -0,0 +1,736 @@
+        # provider is used for which domain so authorizations can be handled gracefully
+        # https://serverfault.com/questions/906407/lets-encrypt-dns-challenge-with-multiple-public-dns-providers
+        verrors = ValidationErrors()
+
+        max_progress = (progress * 4) - progress - (progress * 4 / 5)  # TODO: Refine this
+
+        for authorization_resource in order.authorizations:
+            try:
+                status = False
+                progress += (max_progress / len(order.authorizations))
+                domain = authorization_resource.body.identifier.value  # TODO: handle wildcards
+                # BOULDER DOES NOT RETURN WILDCARDS FOR NOW - WILDCARDS SHOULD BE GRACEFULLY HANDLED IN THE DOMAIN DNS
+                # MAPPING DICT
+                # OTHER IMPLEMENTATIONS RIGHT NOW ASSUME THAT EVERY DOMAIN HAS A WILD CARD IN CASE OF DNS CHALLENGE
+                challenge = None
+                for chg in authorization_resource.body.challenges:
+                    if chg.typ == 'dns-01':
+                        challenge = chg
+
+                if not challenge:
+                    verrors.add(
+                        'acme_authorization.domain',
+                        f'DNS Challenge not found for domain {authorization_resource.body.identifier.value}'
+                    )
+                    continue
+
+                try:
+                    status = self.middleware.call_sync(
+                        'dns.authenticator.update_txt_record', {
+                            'authenticator': domain_names_dns_mapping[domain],
+                            'challenge': challenge.json_dumps(),
+                            'domain': domain,
+                            'key': key.json_dumps()
+                        }
+                    )
+                except ValidationErrors as v:
+                    verrors.extend(v)
+                else:
+                    try:
+                        acme_client.answer_challenge(challenge, challenge.response(key))
+                    except errors.UnexpectedUpdate as e:
+                        verrors.add(
+                            'acme_authorization.domain',
+                            f'Error answering challenge for {domain} : {e}'
+                        )
+
+            finally:
+                job.set_progress(
+                    progress,
+                    f'DNS challenge {"completed" if status else "failed"} for {domain}'
+                )
+
+        if verrors:
+            raise verrors
+
+    @periodic(86400, run_on_start=True)
+    @job(lock='acme_cert_renewal')
+    def renew_certs(self, job):
+        certs = self.middleware.call_sync('acme.query')
+        progress = 0
+        for cert in certs:
+            # TODO: Make the renew date customizable, perhaps add a field to the cert model
+            # TODO: Make renewal optional with default set as True
```

Comment:
Why renew optional?

Reply:
It's a feature. Let's encrypt provides free SSL certs, maybe the user does not want to renew it. So he/she can just set it to False instead of revoking it manually after 3 months. If you think this is not ideal, I'll remove the todo

Reply:
Why would that be the case? The sole purpose of ACME is to make it renew automatically, is there a use-case for this? If renew is not required it seems like the cert could just be deleted.

Reply:
I'll have the todo removed : )